pkg/messaging/v1/invoke_method_request.go

```diff
@@ -165,14 +227,50 @@ func (imr *InvokeMethodRequest) Message() *commonv1pb.InvokeRequest {
 	return imr.r.Message
 }
 
-// RawData returns content_type and byte array body.
-func (imr *InvokeMethodRequest) RawData() (string, []byte) {
+// HasMessageData returns true if the message object contains a slice of data buffered.
+func (imr *InvokeMethodRequest) HasMessageData() bool {
+	m := imr.r.Message
+	return m != nil && m.Data != nil && len(m.Data.Value) > 0
+}
+
+// ContenType returns the content type of the message.
+func (imr *InvokeMethodRequest) ContentType() string {
 	m := imr.r.Message
-	if m == nil || m.Data == nil {
-		return "", nil
+	if m == nil {
+		return ""
 	}
 
-	return m.ContentType, m.Data.Value
+	return m.GetContentType()
+}
+
+// RawData returns the stream body.
+// Note: this method is not safe for concurrent use.
+func (imr *InvokeMethodRequest) RawData() (r io.Reader) {
+	m := imr.r.Message
+	if m == nil {
+		return nil
+	}
+
+	// If the message has a data property, use that
+	if imr.HasMessageData() {
+		return bytes.NewReader(m.Data.Value)
+	}
+
+	return imr.replayableRequest.RawData()
+}
+
+// RawDataFull returns the entire data read from the stream body.
+func (imr *InvokeMethodRequest) RawDataFull() ([]byte, error) {
+	// If the message has a data property, use that
+	if imr.HasMessageData() {
+		return imr.r.Message.Data.Value, nil
```

Comment:
Trying to understand it in a better way: My assumption was that if Data.Value would be beyond a certain threshold size, streaming would be enabled on it. But, it seems that if  Data and its Value is present, then still it would be returned like previously, irrespective of the size.
Or, am I missing something.

Reply:
`data.Value` has data if the message itself (returned by gRPC) has data inside. This is normally happening with gRPC service invocation.

In most cases, the data is set with `imr.WithRawData()` and that always sets a stream.